add some more interfaces to BaseInterfaces.jl

rafaqz opened the pull request:
This PR adds the start of `AbstractSet`, `AbstractDict`, and `AbstractArray` to BaseInterfaces.jl

Hopefully this will end with interface tests that help people write interfaces for these base types and
know they are correct and complete.

@gdalle thoughts or contributions to the cause would be appreciated!

gdalle commented:
Wow, large effort! I'll try to review it in the coming days. Just wondering, did you check out other initiatives like https://github.com/JuliaArrays/ArrayInterface.jl?

rafaqz commented:
Yeah, I've been following ArrayInterface for some years.

But ArrayInterface.jl really isn't the same as the base `AbstractArray` interface. As the docs say: 

> The purpose of this library is to solidify *extensions* to the current AbstractArray interface

(emphasis mine)

The idea in BaseInterfaces.jl is just to have a simple definitions of Base Julia interfaces as far as they are documented, or as used in Base where they are not documented. Its a test of Interfaces.jl capacity to represent a range of interfaces, and should be useful for people extending theses interfaces to have a predefined test suit for it. I'm not sure that the `@implements` declarations will actually be used or useful, but that's ok.

Of course, there will be some overlap in terms of traits as ArrayInterface.jl does include some traits for simple base interface things as well.

gdalle commented:
Feel free to ping me when this is more or less ready for review!

rafaqz commented:
Any time really, feedback would be good. It will be a work in progress untul we register, so we can merge PRs as we go.

gdalle commented:
Okay cool, I'll try to take a look

Seelengrab commented on BaseInterfaces/src/array.jl:
```diff
@@ -0,0 +1,93 @@
+            A -> A[begin] isa eltype(A),
+            A -> A[map(first, axes(A))...] isa eltype(A),
+        ),
+        indexstyle = A -> IndexStyle(A) in (IndexCartesian(), IndexLinear()),
```
This isn't mandatory - there is a default for this.

rafaqz replied:
So it will still always work? Its saying its mandatory tgat it returns one of those options.

Seelengrab commented on BaseInterfaces/src/array.jl:
```diff
@@ -0,0 +1,93 @@
+    # TODO implement all the optional conditions
+    optional = (;
+        setindex! = (
+            A -> length(A) > 1 || throw(ArgumentError("Test arrays must have more than one element to test setindex!")),
```
I think a better property here is to check that setting any index from `eachindex` and retrieving that index again gives the same element. This isn't actually defined by Base though... In some contexts, this can be only temporarily true, e.g. if you have a `DistributedArray` that fails to propagate some information in an overlapping region due to a network failure, for example.

Seelengrab commented on BaseInterfaces/src/dict.jl:
```diff
@@ -0,0 +1,17 @@
+
+@interface DictInterface AbstractDict (
+    mandatory = (;
```
What about `keys` and `values`?

rafaqz replied:
Yeah Dict is not really fleshed out yet

Seelengrab commented on BaseInterfaces/src/set.jl:
```diff
@@ -0,0 +1,60 @@
+            x = first(iterate(s))
+            !in(delete!(s, x), x)
+        end,
+        push! = "push! adds element x to the set" => s -> begin
```
This assumes non-emptiness.

rafaqz replied:
I added a !isempty test in mandatory

Seelengrab commented on BaseInterfaces/src/set.jl:
```diff
@@ -0,0 +1,60 @@
+            s1 = Base.copymutable(s)
+            s1 isa typeof(s) && s1 !== s
+        end,
+        # TODO is there anything we can actually test here?
```
Unfortunately not, since there is no `capacity` function in Base :/

rafaqz replied:
Its so weird that the hint just disappears and we cant access it.

Seelengrab replied:
Well it's just a hint/suggestion, not a requirement :shrug:

Seelengrab commented on BaseInterfaces/src/set.jl:
```diff
@@ -0,0 +1,60 @@
+            s1 = Base.empty(s, Int)
+            eltype(s1) == Int || return false
+        end,
+        emptymutable = "returns a empty set able to hold elements of type U" => s -> begin
```
Also `isempty`

rafaqz replied:
Somehow I forgot aboit `isempty` 😅

Seelengrab commented on BaseInterfaces/src/set.jl:
```diff
@@ -0,0 +1,60 @@
+        intersect = s -> intersect(s, s) == s, # TODO
+        union = s -> union(s, s) == s, # TODO
+        empty! =  "empty! removes all elements from the set" => s -> length(empty!(s)) == 0,
+        delete! = "delete! removes element valued x of the set" => s -> begin
```
This assumes non-emptiness.

rafaqz replied:
Fixed as above by requiring the test data is not empty

Seelengrab commented on BaseInterfaces/src/array.jl:
```diff
@@ -0,0 +1,93 @@
+            A -> eltype(A) isa Type,
+            A -> eltype(A) == _eltype(A),
+        ),
+        ndims = (
```
This should have a fallback definition already, no? IIRC `ndims` comes from having `size` defined already.

rafaqz replied:
Doesnt it use the second type parameter of AbstractArray? 

I just wanted to check that was not messed up in the implementation of a custom type. Maybe not titally useful

Seelengrab replied:
True, but that's not an issue because you have to declare where in a subtype the second parameter of a parent type goes. E.g. it doesn't matter that I can swap the parameters around in my definition:

```
julia> struct MyArr{N,T} <: AbstractArray{T,N} end

julia> ndims(MyArr{2,Int})
2
```

Seelengrab reviewed (commented):
Just some quick things I noticed :)

gdalle reviewed (commented):
I see that @Seelengrab has reviewed the array, dict and set interfaces, so I'll submit my review on everything except those three files for now

rafaqz commented on BaseInterfaces/src/iteration.jl:
```diff
@@ -74,18 +70,19 @@ end
         iterate = test_iterate,
-        size = test_size,
-        eltype = test_eltype,
+        isiterable = x -> Base.isiterable(typeof(x)),
```
Yes its not, but it should be. Its in there to make people implement it in future.

The docs are unfortunately not a reliable spurce of truth.

rafaqz commented on BaseInterfaces/test/runtests.jl:
```diff
@@ -2,10 +2,50 @@ using BaseInterfaces
 using Test
 
-@testset "BaseInterfaces.jl" begin
+@implements SetInterface{(:empty,:emptymutable,:hasfastin,:intersect,:union,:sizehint)} Test.GenericSet
```
I using `subtypes(AbstractSet)` to get everything I can find.

rafaqz commented on BaseInterfaces/src/BaseInterfaces.jl:
```diff
@@ -2,15 +2,43 @@ module BaseInterfaces
 
 # Some example interface delarations.
+
+# @implements ArrayInterface Base.LogicalIndex # No getindex
```
Base just breaks the interface 😱

I have to special case it in DimensionalData.jl for this reason...

rafaqz commented on BaseInterfaces/test/runtests.jl:
```diff
@@ -2,10 +2,50 @@ using BaseInterfaces
-@testset "BaseInterfaces.jl" begin
+@implements SetInterface{(:empty,:emptymutable,:hasfastin,:intersect,:union,:sizehint)} Test.GenericSet
+@implements DictInterface Test.GenericDict
+
```
I dont either :)

rafaqz commented on BaseInterfaces/test/runtests.jl:
```diff
@@ -2,10 +2,50 @@ using BaseInterfaces
+@testset "DictInterface" begin
+    @test Interfaces.test(DictInterface, Dict, [Arguments(d=Dict(:a => 1, :b => 2), k=:c, v=3)])
+    @test Interfaces.test(DictInterface, IdDict, [Arguments(d=IdDict(:a => 1, :b => 2), k=:c, v=3)])
+    @test Interfaces.test(DictInterface, Base.EnvDict, [Arguments(d=Base.EnvDict())])
```
EnvDict fills itself with Env key/values

rafaqz commented on BaseInterfaces/src/iteration.jl:
```diff
@@ -74,18 +70,19 @@ end
     # Mandatory conditions: these must be met by all types
     # that implement the interface.
     mandatory = (
         iterate = test_iterate,
```
Yes often we need to force the test data to have a minimum length. 

But in this case we cant always call length on an iterator so not sure how to test it

rafaqz commented on BaseInterfaces/test/runtests.jl:
```diff
@@ -2,10 +2,50 @@ using BaseInterfaces
+    @test Interfaces.test(ArrayInterface, Base.Slice, [Base.Slice(100:150)])
+    @test Interfaces.test(ArrayInterface, Base.IdentityUnitRange, [Base.IdentityUnitRange(100:150)])
+    @test Interfaces.test(ArrayInterface, Base.CodeUnits, [codeunits("abcde")])
+    # No `getindex` defined for LogicalIndex
```
It only fails with e.g.  `CUDA.allowscalar(true)` on GPU. So thst would need to be `false`.

The docs really specificially say its one of the main mandatory methods, I kinda want the failed test there to demonstrate the problem...

rafaqz commented on BaseInterfaces/src/iteration.jl:
```diff
@@ -74,18 +70,19 @@ end
     # interface type if an object implements them: IterationInterface{(:reverse,:indexing)}
     optional = (
         reverse = test_reverse,
         indexing = test_indexing,
```
I think iterators like that just cant just define `getindex`.

rafaqz commented on BaseInterfaces/src/iteration.jl:
```diff
@@ -49,9 +49,10 @@ function test_size(x)
 end
 
-function test_eltype(x)
+function test_iterator_eltype(x)
```
I refactored all of this to be inline instead, I think having the function outside its harder to understand the interface

rafaqz commented on BaseInterfaces/test/runtests.jl:
```diff
@@ -2,10 +2,50 @@ using BaseInterfaces
+    @test Interfaces.test(SetInterface, BitSet, [BitSet((1, 2))])
+    @test Interfaces.test(SetInterface, Base.KeySet, [Base.KeySet(Dict(:a=>1, :b=>2))])
+    @test Interfaces.test(SetInterface, Test.GenericSet, [Test.GenericSet(Set((1, 2)))])
+    # @test Interfaces.test(SetInterface, Base.IdSet, ?) 
```
Just annoying to construct. Fixed

rafaqz commented on BaseInterfaces/test/runtests.jl:
```diff
@@ -2,10 +2,50 @@ using BaseInterfaces
+    @test Interfaces.test(DictInterface, IdDict, [Arguments(d=IdDict(:a => 1, :b => 2), k=:c, v=3)])
+    @test Interfaces.test(DictInterface, Base.EnvDict, [Arguments(d=Base.EnvDict())])
+    @test Interfaces.test(DictInterface, Base.ImmutableDict, [Arguments(d=Base.ImmutableDict(:a => 1, :b => 2))])
+    # @test Interfaces.test(DictInterface, Base.Pairs, [Arguments(d=Base.pairs((a = 1, b = 2)))])
```
Fixed

rafaqz commented:
Thanks for the feedback, I've implemented most of these and a bunch more. Also fixed up testing and display a lot and accidentally merged it here! But will merge as-is anyway and more can be added in another PR.
